File: ports/rust/src/core/containers/bcsmap.rs
use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use std::hash::Hash;

// BcsMap is a thread-safe generic hash map
pub struct BcsMap<K, V> {
    items: Arc<RwLock<HashMap<K, V>>>,
}

impl<K: Eq + Hash, V: Clone> BcsMap<K, V> {
    pub fn new() -> Self {
        BcsMap {
            items: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub fn insert(&self, key: K, value: V) {
        let mut map = self.items.write().unwrap();
        map.insert(key, value);
    }

    pub fn value(&self, key: &K) -> Option<V> {
        let map = self.items.read().unwrap();
        map.get(key).cloned()
    }

    pub fn contains(&self, key: &K) -> bool {
        let map = self.items.read().unwrap();
        map.contains_key(key)
    }
}

File: ports/go/src/core/containers/bcsmap.go
package containers

import "sync"

// BcsMap is a thread-safe generic hash map
type BcsMap[K comparable, V any] struct {
	mu    sync.RWMutex
	items map[K]V
}

// NewBcsMap creates a new empty map
func NewBcsMap[K comparable, V any]() *BcsMap[K, V] {
	return &BcsMap[K, V]{
		items: make(map[K]V),
	}
}

// Insert adds or updates a key-value pair
func (m *BcsMap[K, V]) Insert(key K, value V) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.items[key] = value
}

// Value retrieves a value by key
func (m *BcsMap[K, V]) Value(key K) (V, bool) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	val, exists := m.items[key]
	return val, exists
}

// Contains checks if a key exists
func (m *BcsMap[K, V]) Contains(key K) bool {
	m.mu.RLock()
	defer m.mu.RUnlock()
	_, exists := m.items[key]
	return exists
}
